### PR TITLE
Add support for more named curves

### LINKED ISSUE #55031: Clients fail to connect to Elasticsearch over https when elasticsearch uses EC private keys and certificates stored in PEM files
**Elasticsearch version** (`bin/elasticsearch --version`): Version: 7.6.1, Build: default/tar/aa751e09be0a5072e8570670309b1f12348f023b/2020-02-29T00:15:25.529771Z, JVM: 13.0.2

**Plugins installed**: []

**JVM version** (`java -version`): JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/13.0.2/13.0.2+8]

**OS version** (`uname -a` if on a Unix-like system): OS[Linux/3.10.0-1062.9.1.el7.x86_64/amd64]

**Description of the problem including expected versus actual behavior**:
Clients ( `curl` , `openssl s_client` ) fail to connect to Elasticsearch when elasticsearch uses EC key material for TLS. Clients fail to complete the TLS handshake

### PR DESCRIPTION
**Pull request by jkakavas**
We implicitly only supported the prime256v1 ( aka secp256r1 )
curve for the EC keys we read as PEM files to be used in any
SSL Context. We would not fail when trying to read a key
pair using a different curve but we would silently assume
that it was using `secp256r1` which would lead to strange
TLS handshake issues if the curve was actually another one.

This commit fixes that behavior in that it
supports parsing EC keys that use any of the named curves
defined in rfc5915 and rfc5480 making no assumptions about
whether the security provider in use supports them (JDK8 and
higher support all the curves defined in rfc5480).

resolves: #55031

### CODE REVIEW

File: libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemUtils.java [hunk 1]
```diff
@@ -602,4 +605,42 @@ private static String getKeyAlgorithmIdentifier(byte[] keyBytes) throws IOExcept
...
+                return "sect409r1";
+            case "1.3.132.0.35":
+                return "secp521r1";
+            case "1.3.132.0.38":
+                return "sect571k1";
+            case "1.3.132.0.39":
+                return "sect571r1";
+        }
+        throw new GeneralSecurityException("Error parsing EC named curve identifier. Named curve with OID: " + oidString + " is not " +
+            "supported");
```
Comment (tvernum) [inline libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/PemUtils.java:643]:
```suggestion
        throw new GeneralSecurityException("Error parsing EC named curve identifier. Named curve with OID: " + oidString 
            + " is not supported");
```

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by ywangd** [comment]
I am trying to understand this issue better. Could you please help with following questions? Thanks!

* This was not an issue for keystore files because `KeyStoreUtil` basically wraps JDK APIs to load the file and handles curves correctly?
* Why are there two sets of PemUtils and related readme and the changes have to be added into both of them?

**Comment by tvernum** [comment]
This was not an issue for Keystore files because the bug was in our reading of PEM formatted keys. The JDK does not ship with code to read PEM formatted private keys. `CertifcateFactory` can read PEM certificates, but there is nothing for keys.

Historically we used BouncyCastle for key reading, but we had to stop shipping BC as a runtime dependency so that we could support FIPS mode JVMs that use BC-FIPS. There were classpath issues if we used a JVM with BC-FIPS as its JCE provider and we added regular BC onto the classpath.
So we (@jkakavas) wrote our own PEM key parser which, as it turns out, had a bug in that it assumed a single curve for EC keys.

Keystore has support for reading JKS and PKCS#12 keys, which means there's probably a bunch of code in the JRE (or more accurate the Sun/Oracle JCE provider) that could have been useful for us when decoding key/curve identifiers etc, but it's not exposed anywhere we can make use of, so we have our own.

We have two SSL configuration implementations in ES. One is opensource and is used by reindex, the other is Elastic licensed and is used within X-Pack. The Elastic licensed one predates the OSS one. When we (@tvernum) added the OSS one we based it on the existing Elastic code, but that code was tightly coupled to other Elastic licensed pieces, so it wasn't simple to extract cleanly. That meant that we wrote a new implementation based on the existing code, and took the oppportunity to clean some of it up in the process. However, for both BWC and progress-over-perfection reasons we didn't switch the X-Pack code to use the new implementations at that time. I would like to get rid of the X-Pack implementation in ES8.0, but I know there are small differences between the two implementations, so I don't think we can safely swap it out in a minor release.

**Comment by ywangd** [comment]
Thanks @tvernum for the detailed explanation.

**Pull request merged**